### DISCUSSION TITLE
Gitleaks is not detecting passwords

### DISCUSSION THREAD

**Discussion by sophallee** [zricethezav/gitleaks#1399] [General]
I'm having trouble getting Gitleaks to detect known passwords in my test commits.

I installed Gitleaks using the following script "bash <(curl https://gitlab.com/gitlab-com/gl-security/security-research/gitleaks-endpoint-installer/-/raw/main/install.sh)". This script created a folder  ~/.gitlab-gitleaks which contains the Gitleaks executable.

Test Scenario:
- I'm running test commits with passwords stored in a file.
- However, Gitleaks doesn't report any leaks

Steps to Reproduce:
- git commit -m "Test" (This commit includes the test file with passwords)
- Gitleaks output:
```
10:10PM INF scan completed in 57.565738ms
10:10PM INF no leaks found
```

I've also tried adding a .gitleaks.toml file in the root directory (content not shown for security reasons), but it doesn't seem to make a difference.

Test File Content:
```
PASSWORD='a123qs#$233a'
password='62sa$23das'
```
Pre-commit Hhook Content ( ~/.gitlab-gitleaks/hooks/pre-commit)
```
#!/bin/bash

if [[ -n $I_WANT_GITLEAKS_SKIP ]]
  then
  exit 0
fi

MSG="\n\n\033[31mCommit failed.\n\n\033[0mgitleaks has detected some secrets in your commit.\nPlease see https://about.gitlab.com/handbook/engineering/security/gitleaks.html for more details."
$HOME/.gitlab-gitleaks/gitleaks protect --staged --redact --verbose || (printf "${MSG}" ; exit -1)
```
Explicitly setting the gitleaks.toml with --config <PATH> does not help. Attached is the gitleaks.toml. 

Do I need to configure Gitleaks further for it to detect these specific password formats? Thanks
[gitleaks.txt](https://github.com/gitleaks/gitleaks/files/15189034/gitleaks.txt)

**Comment by sophallee** [comment]
I figured out the issue! Gitleaks uses a default rule set that might not cover all secret formats by default. Here's how to address this:

Create or edit your `.gitleaks.toml` file and add the following under the `[extend]` section:
```
[extend]
useDefault = false
```
This tells Gitleaks to not use the default rules and rely on your custom configuration.

If you want Gitleaks to run as a pre-commit hook (automatically checking for leaks before committing), you can set an environment variable pointing to your `pre-commit` hook. 
```
GITLEAKS_CONFIG="$HOME/.gitlab-gitleaks/config/gitleaks.toml"
export GITLEAKS_CONFIG
```

**Comment by sophallee** [comment]
I discovered a possible bug in Gitleaks' behavior. After adding a password entry to a test file, removing it, and then adding it again, Gitleaks fails to detect the leak during subsequent scans. This suggests that Gitleaks might be caching some information about previously detected secrets.

Has anyone else encountered this behavior?

 I'm curious if others have experienced this persistence issue. If so, is there a known solution or workaround besides disabling the default rules?

**Reply by bplaxco** [reply]
Just curious, is this the case with the most recent versions of gitleaks? (It looks like the GitLab installer might be a bit dated).
